### PR TITLE
Allow setting target allocator via label

### PR DESCRIPTION
**Pull request by swiatekm**
**Description:**
Allow connecting the OpenTelemetryCollector and TargetAllocator CRs by setting a label on the Collector CR.

**Link to tracking Issue(s):** <Issue number if applicable>

- Related to: #2422 

**Testing:**
Added some basic unit tests and an e2e test showing how label changes affect the Collector manifests.

### CODE REVIEW

File: controllers/opentelemetrycollector_controller.go [hunk 1]
```diff
@@ -50,6 +51,8 @@ import (
 	"github.com/open-telemetry/opentelemetry-operator/pkg/featuregate"
 )
 
+const collectorTargetAllocatorLabelName = "opentelemetry.io/target-allocator"
```
Comment (jaronoff97) [inline controllers/opentelemetrycollector_controller.go:54]: should we put this [here](https://github.com/open-telemetry/opentelemetry-operator/blob/main/pkg/constants/env.go)?

File: controllers/opentelemetrycollector_controller.go [hunk 2]
```diff
@@ -179,14 +182,28 @@ func (r *OpenTelemetryCollectorReconciler) GetParams(instance v1beta1.OpenTeleme
...
+	targetAllocator, err := r.getTargetAllocator(ctx, p)
 	if err != nil {
 		return p, err
 	}
 	p.TargetAllocator = targetAllocator
 	return p, nil
 }
 
+func (r *OpenTelemetryCollectorReconciler) getTargetAllocator(ctx context.Context, params manifests.Params) (*v1alpha1.TargetAllocator, error) {
+	otelcol := params.OtelCol
```
Comment (jaronoff97) [inline controllers/opentelemetrycollector_controller.go:194]: why do we need the shallow copy here?

Comment (swiatekm) [reply controllers/opentelemetrycollector_controller.go:194]: This was just for readability I think, but it doesn't look necessary anymore. It's a pointer anyway, so no copy.

File: main.go [hunk 3]
```diff
@@ -440,7 +440,7 @@ func main() {
 
 		bv := func(collector otelv1beta1.OpenTelemetryCollector) admission.Warnings {
 			var warnings admission.Warnings
-			params, newErr := collectorReconciler.GetParams(collector)
+			params, newErr := collectorReconciler.GetParams(context.Background(), collector)
```
Comment (frzifus) [inline main.go:443]: Does the underlying implementation handle timeouts and all that stuff?

**Review by frzifus — APPROVED** [review]
:partying_face: :partying_face:

File: main.go (hunk 3 above)
Comment (swiatekm) [reply main.go:443]: It does, but now that you mention it, this should really be the context in which the caller runs. I'll modify the build validation function itself to take a context argument instead.

**Review by jaronoff97 — APPROVED** [review]

File: main.go (hunk 3 above)
Comment (swiatekm) [reply main.go]: Fixed.

**Pull request merged**